perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/MembersManagerBlImpl.java

```diff
@@ -2328,34 +2322,58 @@ public Member createSponsoredMember(PerunSession session, Vo vo, String namespac
-		p.put(PasswordManagerModule.PASSWORD_KEY,password);
-		Map<String, String> r = getPerunBl().getUsersManagerBl().generateAccount(session, namespace, p);
-		String login = r.get(loginAttributeName);
-		setLoginToSponsoredUser(session,sponsoredUser,loginAttributeName,login);
+		return setSponsoredMember(session, vo, sponsoredUser, namespace, password, sponsor, asyncValidation);
+	}
+
+	@Override
+	public Member setSponsoredMember(PerunSession session, Vo vo, User userToBeSponsored, String namespace, String password, User sponsor, boolean asyncValidation) throws AlreadyMemberException, ExtendMembershipException, UserNotInRoleException, WrongAttributeValueException, WrongReferenceAttributeValueException, LoginNotExistsException, PasswordCreationFailedException, InvalidLoginException, ExtSourceNotExistsException {
+		//check that sponsoring user has role SPONSOR for the VO
+		if (!getPerunBl().getVosManagerBl().isUserInRoleForVo(session, sponsor, Role.SPONSOR, vo, true)) {
+			throw new UserNotInRoleException("user " + sponsor.getId() + " is not in role SPONSOR for VO " + vo.getId());
+		}
+		String loginAttributeName = PasswordManagerModule.LOGIN_PREFIX + namespace;
+		String attributeValue = getAttributeValueAsString (session, userToBeSponsored, loginAttributeName);
+
+		//create the user account in external system if does not exist yet
+		String login = createUserAccountInExternalSystem (session, userToBeSponsored, namespace, attributeValue, loginAttributeName, password);
 
 		//create the member in Perun
-		Member sponsoredMember = getMembersManagerImpl().createSponsoredMember(session, vo, sponsoredUser, sponsor);
+		Member sponsoredMember = getMembersManagerImpl().createSponsoredMember(session, vo, userToBeSponsored, sponsor);
 		getPerunBl().getAuditer().log(session, new MemberCreated(sponsoredMember));
 		getPerunBl().getAuditer().log(session, new SponsoredMemberSet(sponsoredMember));
 		getPerunBl().getAuditer().log(session, new SponsorshipEstablished(sponsoredMember, sponsor));
 		extendMembership(session, sponsoredMember);
 		insertToMemberGroup(session, sponsoredMember, vo);
-		if(asyncValidation) {
+
+		if (asyncValidation) {
 			validateMemberAsync(session, sponsoredMember);
 		} else {
 			//for unit tests
 			validateMember(session, sponsoredMember);
 		}
-		getPerunBl().getUsersManagerBl().validatePasswordAndSetExtSources(session, sponsoredUser, login, namespace);
+		getPerunBl().getUsersManagerBl().validatePasswordAndSetExtSources(session, userToBeSponsored, login, namespace);
+
 		return sponsoredMember;
 	}
 
+	/**
+	 * Try to get attribute from attribute manager
+	 *
+	 * @param session perun session
+	 * @param userToBeSponsored user, that will be sponsored by sponsor
+	 * @param loginAttributeName login attribute name
+	 * @return attributeValue
+	 */
+	private String getAttributeValueAsString (PerunSession session, User userToBeSponsored, String loginAttributeName) {
```

Comment:
Sorry, I did not write the comment about "moving the try-catch block to the new method" properly. I did not mean to create a new private method with this code. I meant to move this block of code to the new private method `createUserAccountInExternalSystem`, which you have already created. You can move also the whole variable `loginAttributeName` to that method as you will need it only there. Of course, the private method `createUserAccountInExternalSystem` will need fewer parameters, because of that, and the method `getAttributeValueAsString` can be deleted.

Reply:
However, it is not a big deal. We can leave it as it is and refactor it in the future if there will be a need for that.